### ISSUE TITLE
"GeometryError" when runing eudemo/reactor.py

### ISSUE THREAD

**Issue by afshin-arj** [Fusion-Power-Plant-Framework/bluemira#1828]
## Describe the bug

I tried to run the example for EU-DEMO with default parameters (../eudemo/eudemo/reactor.py) and received the following error. I appreciate for any recommendations in advance.  

## Steps to reproduce

python eudemo/reactor.py

--->
```
+-------------------------------------------------------------------------+
| WARNING: PROCESS not installed on this machine; cannot run PROCESS.     |
+-------------------------------------------------------------------------+
+-------------------------------------------------------------------------+
| Mocking PROCESS systems code run                                        |
+-------------------------------------------------------------------------+
+-------------------------------------------------------------------------+
| WARNING: BoundedVariable 'x1': value was set to above its upper         |
| bound. Adjusting bound.                                                 |
+-------------------------------------------------------------------------+
+-------------------------------------------------------------------------+
| WARNING: BoundedVariable 'height': value was set to below its lower     |
| bound. Adjusting bound.                                                 |
+-------------------------------------------------------------------------+
+-------------------------------------------------------------------------+
| Solving design problem: MinimiseLengthGOP                               |
+-------------------------------------------------------------------------+
+-------------------------------------------------------------------------+
| WARNING:                                                                |
| NLopt Optimiser was halted due to round-off errors. A useful result     |
| was probably found...                                                   |
+-------------------------------------------------------------------------+
+-------------------------------------------------------------------------+
| WARNING: make_polygon: input points are closed but closed=False,        |
| defaulting to closed=True.                                              |
+-------------------------------------------------------------------------+
Traceback (most recent call last):
  File "eudemo/reactor.py", line 161, in <module>
    reactor.tf_coils = build_tf_coils(
  File "eudemo/reactor.py", line 117, in build_tf_coils
    return TFCoil(builder.build(), builder._make_field_solver())
  File "/home/afshin/Software/fusion/bluemira/eudemo/eudemo/tf_coils.py", line 427, in build
    xyz_case, xyz = self.build_xyz(y_in, ins_inner_face, ib_cas_wire, ob_cas_wire)
  File "/home/afshin/Software/fusion/bluemira/eudemo/eudemo/tf_coils.py", line 494, in build_xyz
    case_solid, case_sector = self._build_xyz_case(
  File "/home/afshin/Software/fusion/bluemira/eudemo/eudemo/tf_coils.py", line 664, in _build_xyz_case
    pieces = boolean_cut(solid, inboard_casing)
  File "/home/afshin/Software/fusion/bluemira/bluemira/geometry/tools.py", line 1165, in boolean_cut
    return [convert(obj, shape.label) for obj in cut_shape]
  File "/home/afshin/Software/fusion/bluemira/bluemira/geometry/tools.py", line 1165, in <listcomp>
    return [convert(obj, shape.label) for obj in cut_shape]
  File "/home/afshin/Software/fusion/bluemira/bluemira/codes/_freecadapi.py", line 84, in wrapper
    return func(*args, **kwargs)
  File "/home/afshin/Software/fusion/bluemira/bluemira/geometry/tools.py", line 64, in convert
    output = BluemiraSolid._create(apiobj, label)
  File "/home/afshin/Software/fusion/bluemira/bluemira/geometry/solid.py", line 82, in _create
    raise GeometryError(f"Solid {obj} is not valid.")
bluemira.geometry.error.GeometryError: Solid <Solid object at 0x55a77af85d30> is not valid.
```

**Comment by je-cook** [comment]
Hi @afshin-arj, thanks for trying bluemira!

Unfortunately this is a known bug (although admittedly not explicitly open). The EU-DEMO workflow is still being built out and #309 is tracking this along with everything under the [EU-DEMO label](https://github.com/Fusion-Power-Plant-Framework/bluemira/issues?q=is%3Aissue+is%3Aopen+label%3AEU-DEMO).

If you comment out L153-L159 (the section where your error comes from) and L168-L175 (also has a bug) of `reactor.py` you should be able to run the script see the reactor CAD. I recommend running with interactive mode because then you will be able to inspect the reactor object after the run: `python -i eudemo/reactor.py`

We aim to finish off the EU-DEMO workflow over the next couple of months. Our examples folder has some more working examples if you need.

Btw I've just reformatted your stack trace with triple backticks so its easier to read

**Comment by je-cook** [comment]
I should probably also note that a lot of `GeometryError`s are down to #1347 and FreeCAD, we have some workarounds but some errors will sporadically appear until that fix is in.

If there is a specific usecase of bluemira you are interested in please let us know, we may be able provide some support if needed.

**Comment by afshin-arj** [comment]
Hi @je-cook for your kind answer. 
Based on your recommendations, with commenting out the lines related to vacuum vessel and PF/TF coils, no GeometryError popped up. But as I recall, in early versions of bluemira, all parts were successfully simulated and a complete CAD output was generated.
Also, our team has recently initiated conceptual design of a model spherical tokamak (similar to STEP); So I've tried to use bluemira in the past few months to get a rough estimation of the model spherical tokamak with simply changing aspect ratio and number of coils in EUDEMO example, which all failed unfortunately.

**Comment by je-cook** [comment]
Quite a few things have changed in the last couple of months with the aim of making bluemira a bit more flexible. It has meant that we have needed to rewrite a lot of functionality in the short term in the hope that everything is more maintainable long term.

We've split more explicitly the difference between a reactor design that uses bluemira as a framework and the framework itself (See [the new architecture](https://bluemira.readthedocs.io/en/latest/introduction.html#high-level-architecture) and [the user-designer-developer split](https://bluemira.readthedocs.io/en/latest/introduction.html#how-to-use-bluemira)). This does mean that we expect most users to use eudemo as inspiration but may have to adjust the designers and builders to suit their usecase.

Our default PROCESS input and mocked output both assume a DEMO like reactor as is our main usecase. By default we run PROCESS in mock mode which ensures that the example is still runnable without PROCESS installed. To make it usable for more spherical tokamaks you will have to change the output either by editing the mockPROCESS.json to modify the mock output, specifying a read mode with a known good process output, or run PROCESS with some known good inputs.

Although the framework is tentatively v1, the eudemo model is still very much work in progress, and the coils both PF and TF still need a few fixes before they are robust enough to be customiseable directly from the example. Mostly we run into CAD issues; `GeometryError` or even lower level `BRep_API: command not done` are annoyingly common. #1257 hopefully will deal with some of the TF coil issues whereas the PF coils are held up by getting the fixed-free equilibrium solve working in #1733.

Once those couple of issues are complete we hope that everything becomes a bit more useable, please bear with us.

May I ask what branch you are using `develop`, `main` or a specific release (it should have been asked for in the issue template)? 
We recommend sticking to either `main` or a release as we can't ensure that develop will be stable at all times although that should be the case for each release.

**Comment by CoronelBuendia** [comment]
Hi @afshin-arj, we've had a pretty significant overhaul of that detailed example (presently dubbed EU-DEMO). This error no longer crops up where it did.

You should now be able to build the following on `develop` by just running the example file as before:

![image](https://github.com/Fusion-Power-Plant-Framework/bluemira/assets/26097289/06964c61-e88b-4235-98e9-781de3ea3028)

I'm closing this issue, but please feel free to re-open (or open a new one) if you run into any problems.
